File: src/psm-core/src/grounding-guards.ts
import { routeForAction } from "./actions.js";
import { tokenize } from "./ranking.js";
import type { StorageDecision } from "./types.js";

const BLEED_PATTERN = /checkpoint|powershell|gate datasets|nvidia-smi|direct probe|token budget|runpod|fact parser|malformed parser|constoursated|gate6|expanded probe|gate-?\d/i;

export type StorageGuardRejectRoute = "grounding_reject" | "grounding_reject_bleed";

export interface StorageGuardResult {
  decision: StorageDecision;
  rejected: boolean;
  guard_route?: StorageGuardRejectRoute;
  guard_reason?: string;
}

export function hasCurriculumBleed(text: string): boolean {
  return BLEED_PATTERN.test(text);
}

export function significantTokens(text: string): string[] {
  return tokenize(text).filter((token) => token.length >= 3 && !/^\d+$/.test(token));
}

export function groundingOverlapScore(rememberTarget: string, storedText: string): { overlap: number; required: number; grounded: boolean } {
  const inputTokens = significantTokens(rememberTarget);
  if (inputTokens.length === 0) {
    return { overlap: 0, required: 0, grounded: true };
  }
  const storedSet = new Set(significantTokens(storedText));
  const overlap = inputTokens.filter((token) => storedSet.has(token)).length;
  const required = Math.min(2, Math.max(1, Math.ceil(inputTokens.length * 0.1)));
  return { overlap, required, grounded: overlap >= required };
}

export function isGroundedInSource(rememberTarget: string, storedText: string): boolean {
  return groundingOverlapScore(rememberTarget, storedText).grounded;
}

function storedTextFromDecision(decision: StorageDecision): string {
  const content = decision.memory?.content?.trim() ?? "";
  const factParts = (decision.facts ?? []).flatMap((fact) => [
    fact.subject,
    fact.predicate,
    fact.value_text,
    fact.evidence_text ?? ""
  ]);
  return [content, ...factParts].filter(Boolean).join(" ");
}

function wouldPersist(decision: StorageDecision): boolean {
  if (decision.parse_error) return false;
  const route = routeForAction(decision.action);
  if (route === "ignore" || route === "recall_only") return false;
  return Boolean(storedTextFromDecision(decision).trim());
}

export function applyStorageGuards(rememberTarget: string, decision: StorageDecision): StorageGuardResult {
  if (!wouldPersist(decision)) {
    return { decision, rejected: false };
  }
  const storedText = storedTextFromDecision(decision);
  if (hasCurriculumBleed(storedText)) {
    return {
      decision,
      rejected: true,
      guard_route: "grounding_reject_bleed",
      guard_reason: "Stored content matches curriculum bleed blocklist."
    };
  }
  if (!isGroundedInSource(rememberTarget, storedText)) {
    return {
      decision,
      rejected: true,
      guard_route: "grounding_reject",
      guard_reason: "Stored content is not grounded in remember_target tokens."
    };
  }
  return { decision, rejected: false };
}

export function isFailSafeIgnore(result: Record<string, unknown>): boolean {
  const route = typeof result.route === "string" ? result.route : "";
  const reasoning = typeof result.reasoning === "string" ? result.reasoning : "";
  if (route === "parse_error_noop") return true;
  return /model output unparseable|storing nothing/i.test(reasoning);
}

File: dotnet/src/PsmMemory.Core/GroundingGuards.cs
using System.Text.RegularExpressions;
using PsmMemory.Core.Models;

namespace PsmMemory.Core;

public sealed class GroundingOverlapScore
{
    public required int Overlap { get; init; }
    public required int Required { get; init; }
    public required bool Grounded { get; init; }
}

public sealed class StorageGuardResult
{
    public required StorageDecision Decision { get; init; }
    public required bool Rejected { get; init; }
    public string? GuardRoute { get; init; }
    public string? GuardReason { get; init; }
}

/// <summary>
/// Ported directly from psm-core/src/grounding-guards.ts: applyStorageGuards(), hasCurriculumBleed(),
/// groundingOverlapScore(). Rejects (downgrades to ignore) any storage decision whose content isn't
/// grounded in the source text, or that matches the curriculum-bleed blocklist.
/// </summary>
public static partial class GroundingGuards
{
    [GeneratedRegex(
        "checkpoint|powershell|gate datasets|nvidia-smi|direct probe|token budget|runpod|fact parser|malformed parser|constoursated|gate6|expanded probe|gate-?\\d",
        RegexOptions.IgnoreCase)]
    private static partial Regex BleedPattern();

    [GeneratedRegex(@"^\d+$")]
    private static partial Regex AllDigitsRegex();

    public static bool HasCurriculumBleed(string text) => BleedPattern().IsMatch(text);

    public static List<string> SignificantTokens(string text) =>
        Ranking.Tokenize(text).Where(token => token.Length >= 3 && !AllDigitsRegex().IsMatch(token)).ToList();

    public static GroundingOverlapScore GroundingOverlapScore(string rememberTarget, string storedText)
    {
        var inputTokens = SignificantTokens(rememberTarget);
        if (inputTokens.Count == 0)
        {
            return new GroundingOverlapScore { Overlap = 0, Required = 0, Grounded = true };
        }
        var storedSet = new HashSet<string>(SignificantTokens(storedText));
        var overlap = inputTokens.Count(t => storedSet.Contains(t));
        var required = Math.Min(2, Math.Max(1, (int)Math.Ceiling(inputTokens.Count * 0.1)));
        return new GroundingOverlapScore { Overlap = overlap, Required = required, Grounded = overlap >= required };
    }

    public static bool IsGroundedInSource(string rememberTarget, string storedText) =>
        GroundingOverlapScore(rememberTarget, storedText).Grounded;

    private static string StoredTextFromDecision(StorageDecision decision)
    {
        var content = decision.Memory?.Content?.Trim() ?? "";
        var factParts = decision.Facts.SelectMany(fact => new[]
        {
            fact.Subject ?? "",
            fact.Predicate ?? "",
            fact.ValueText ?? "",
            fact.EvidenceText ?? ""
        });
        return string.Join(" ", new[] { content }.Concat(factParts).Where(s => !string.IsNullOrEmpty(s)));
    }

    private static bool WouldPersist(StorageDecision decision)
    {
        if (decision.ParseError is not null) return false;
        var route = Actions.RouteForAction(decision.Action);
        if (route is Actions.Routes.Ignore or Actions.Routes.RecallOnly) return false;
        return !string.IsNullOrWhiteSpace(StoredTextFromDecision(decision));
    }

    public static StorageGuardResult ApplyStorageGuards(string rememberTarget, StorageDecision decision)
    {
        if (!WouldPersist(decision))
        {
            return new StorageGuardResult { Decision = decision, Rejected = false };
        }
        var storedText = StoredTextFromDecision(decision);
        if (HasCurriculumBleed(storedText))
        {
            return new StorageGuardResult
            {
                Decision = decision,
                Rejected = true,
                GuardRoute = "grounding_reject_bleed",
                GuardReason = "Stored content matches curriculum bleed blocklist."
            };
        }
        if (!IsGroundedInSource(rememberTarget, storedText))
        {
            return new StorageGuardResult
            {
                Decision = decision,
                Rejected = true,
                GuardRoute = "grounding_reject",
                GuardReason = "Stored content is not grounded in remember_target tokens."
            };
        }
        return new StorageGuardResult { Decision = decision, Rejected = false };
    }
}
